File: Module 01/Fibonacci.Fs/Fibonacci.fs
namespace Fibonacci.Fs

open System.Runtime.CompilerServices

module Fibonacci =
    let private memoize fn =
      let cache = new System.Collections.Generic.Dictionary<_,_>()
      (fun x ->
        match cache.TryGetValue x with
        | true, v -> v
        | false, _ -> let v = fn (x)
                      cache.Add(x,v)
                      v)
    [<MethodImpl(MethodImplOptions.NoInlining)>]
    let fib n = 
        let rec r_fib x prev curr =
            match x with
               | 2UL -> prev + curr
               | _ ->
                   r_fib (x - 1UL) curr (prev + curr)

        match n with
            | 0UL -> 0UL
            | 1UL -> 1UL
            | _ ->  r_fib n 0UL 1UL

    let memoized =
        memoize fib

File: Module 01/Dotnetos.AsyncExpert.Homework.Module01.Benchmark/Fibonacci.cs
﻿using System.Collections.Generic;
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

namespace Dotnetos.AsyncExpert.Homework.Module01.Benchmark
{
    [DisassemblyDiagnoser(exportCombinedDisassemblyReport: true, maxDepth: 2, printInstructionAddresses: true, exportHtml: true)]
    [MemoryDiagnoser]
    [ShortRunJob]
    public class FibonacciCalc
    {
        // HOMEWORK:
        // ✔ 1. Write implementations for RecursiveWithMemoization and Iterative solutions
        // ✔ 2. Add memory profiler to the benchmark
        // ✔ 3. Run with release configuration and compare results 
        // ✔ 4. Open disassembler report and compare machine code
        // 
        // You can use the discussion panel to compare your results with other students

        [Benchmark]
        [ArgumentsSource(nameof(Data))]
        public ulong Recursive(ulong n)
        {
            if (n == 0) return 0;
            if (n == 1 || n == 2) return 1;
            return Recursive(n - 2) + Recursive(n - 1);
        }

        [Benchmark]
        [ArgumentsSource(nameof(Data))]
        public ulong TailRecursive(ulong n) => 
            Fibonacci.Fs.Fibonacci.fib(n);

        [Benchmark(Baseline = true)]
        [ArgumentsSource(nameof(Data))]
        public ulong Iterative(ulong n)
        {
            return Compute(n);
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        private static ulong Compute(ulong n)
        {
            if (n == 0) return 0;
            ulong head = 1;
            ulong tail = 1;
            for (ulong i = 2; i < n; i++)
            {
                ulong carry = head;
                head = tail;
                tail += carry;
            }
            return tail;
        }

        [Benchmark]
        [ArgumentsSource(nameof(Data))]
        public ulong TailRecursiveMemoized(ulong n) =>
            Fibonacci.Fs.Fibonacci.memoized.Invoke(n);

        public IEnumerable<ulong> Data()
        {
            yield return 15;
            yield return 35;
        }
    }
}
